src/Compilers/Core/CodeAnalysisTest/Collections/DebuggerAttributes.cs

```diff
@@ -0,0 +1,247 @@
+        {
+            // Get the DebuggerDisplayAttribute for obj
+            var objType = obj.GetType();
+            var attrs =
+                objType.GetTypeInfo().CustomAttributes
+                .Where(a => a.AttributeType == typeof(DebuggerDisplayAttribute))
+                .ToArray();
+            if (attrs.Length != 1)
+            {
+                throw new InvalidOperationException($"Expected one DebuggerDisplayAttribute on {objType}.");
+            }
+            var cad = attrs[0];
+
+            // Get the text of the DebuggerDisplayAttribute
+            string attrText = (string)cad.ConstructorArguments[0].Value;
+
+            var segments = attrText.Split(new[] { '{', '}' });
+
+            if (segments.Length % 2 == 0)
+            {
+                throw new InvalidOperationException($"The DebuggerDisplayAttribute for {objType} lacks a closing brace.");
+            }
+
+            if (segments.Length == 1)
+            {
+                throw new InvalidOperationException($"The DebuggerDisplayAttribute for {objType} doesn't reference any expressions.");
+            }
+
+            var sb = new StringBuilder();
+
+            for (int i = 0; i < segments.Length; i += 2)
+            {
+                string literal = segments[i];
+                sb.Append(literal);
+
+                if (i + 1 < segments.Length)
+                {
+                    string reference = segments[i + 1];
+                    bool noQuotes = reference.EndsWith(",nq");
+
+                    reference = reference.Replace(",nq", string.Empty);
+
+                    // Evaluate the reference.
+                    object member;
+                    if (!TryEvaluateReference(obj, reference, out member))
+                    {
+                        throw new InvalidOperationException($"The DebuggerDisplayAttribute for {objType} contains the expression \"{reference}\".");
+                    }
+
+                    string memberString = GetDebuggerMemberString(member, noQuotes);
+
+                    sb.Append(memberString);
+                }
+            }
+
+            return sb.ToString();
+        }
+
+        private static string GetDebuggerMemberString(object member, bool noQuotes)
+        {
+            string memberString = "null";
+            if (member != null)
+            {
+                memberString = member.ToString();
+                if (member is string)
+                {
+                    if (!noQuotes)
+                    {
+                        memberString = '"' + memberString + '"';
+                    }
+                }
+                else if (!IsPrimitiveType(member))
+                {
+                    memberString = '{' + memberString + '}';
+                }
+            }
+
+            return memberString;
+        }
+
+        private static bool IsPrimitiveType(object obj) =>
+            obj is byte || obj is sbyte ||
```

Comment:
Nit: could use `or`s here.

Reply:
This is copied from runtime code per the comment at the top of the file.